### ISSUE TITLE
Unintended behavior when set change states before add to the window

### ISSUE THREAD

**Issue by ReDetection** [AndreasVerhoeven/AvePurchaseButton#5]
First of all, I would like to thank you for this great button and the work you've done to make it look like the AppStore's one! And also for the IBDesignable support, this is really cool! 👍

One thing I noticed that changing your example usage like this causes some glitches:

``` obj-c
- (void)addPurchaseButton {
  AvePurchaseButton* button = [[AvePurchaseButton alloc] initWithFrame:CGRectZero];
    [button addTarget:self action:@selector(purchaseButtonTapped:) forControlEvents:UIControlEventTouchUpInside];
    button.buttonState = AvePurchaseButtonStateProgress;
    button.normalTitle = @"$ 2.99";
    button.confirmationTitle = @"BUY";
    [button sizeToFit];
    [self.view addSubview:button];
}
```

The first glitch is CoreGraphics warnings:

```
     <Error>: CGContextSetBlendMode: invalid context 0x0. Backtrace:
      <UIRectFillUsingBlendMode+180>
       <-[AvePurchaseActivityIndicatorView imageForAnimation]+491>
        <-[AvePurchaseActivityIndicatorView ensureImageViewImage]+129>
         <-[AvePurchaseActivityIndicatorView startAnimating]+199>
          <-[AvePurchaseButton setButtonState:animated:]+703>
           <-[AvePurchaseButton setButtonState:]+64>
```

Seconds is wrong dimensions of activity indicator:
![screen shot 2016-09-28 at 10 34 18](https://cloud.githubusercontent.com/assets/458005/18899949/0abcee70-8567-11e6-86a5-2b68ca62fd91.png) ![screen shot 2016-09-28 at 10 34 37](https://cloud.githubusercontent.com/assets/458005/18899951/0e87ad4c-8567-11e6-9229-00190dec3e6e.png)

**Comment by AndreasVerhoeven** [comment]
Good catch! Should be fixed in version 1.0.4. Thanks for reporting!

**Comment by ReDetection** [comment]
Yep, confirm it's working. Thanks!

**Comment by jonbrown21** [comment]
I am getting some similar issues I have tried 1.0.4 and 1.0.5

```
<Error>: CGContextSetFillColorWithColor: invalid context 0x0. Backtrace:
	  <-[AveBorderedButton fillViewImage]+199>
	   <-[AveBorderedButton updateFillView]+81>
	    <-[AveBorderedButton updateHighlightedStateAnimated]+43>
	     <-[AveBorderedButton setHighlighted:]+88>
	      <-[AvePurchaseButton setHighlighted:]+173>
	       <-[UITableViewCell _updateHighlightColorsForView:highlighted:]+771>
	        <-[UITableViewCell _updateHighlightColorsForView:highlighted:]+1120>
	         <-[UITableViewCell _updateHighlightColors]+115>
	          <-[UITableViewCell prepareForReuse]+421>
	           <-[UITableView _dequeueReusableViewOfType:withIdentifier:]+1476>
	            <-[BFMasterTableViewController tableView:cellForRowAtIndexPath:]+436>
	             <-[UITableView _createPreparedCellForGlobalRow:withIndexPath:willDisplay:]+757>
	              <-[UITableView _createPreparedCellForGlobalRow:willDisplay:]+74>
	               <-[UITableView _updateVisibleCellsNow:isRecursive:]+3295>
	                <-[UITableView _performWithCachedTraitCollection:]+110>
	                 <-[UITableView layoutSubviews]+222>
	                  <-[UIView(CALayerDelegate) layoutSublayersOfLayer:]+1237>
	                   <-[CALayer layoutSublayers]+146>
	                    <_ZN2CA5Layer16layout_if_neededEPNS_11TransactionE+366>
	                     <_ZN2CA5Layer28layout_and_display_if_neededEPNS_11TransactionE+24>
	                      <_ZN2CA7Context18commit_transactionEPNS_11TransactionE+280>
	                       <_ZN2CA11Transaction6commitEv+475>
	                        <_ZN2CA11Transaction17observer_callbackEP19__CFRunLoopObservermPv+113>
	                         <__CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__+23>
	                          <__CFRunLoopDoObservers+391>
	                           <__CFRunLoopRun+1198>
	                            <CFRunLoopRunSpecific+420>
	                             <GSEventRunModal+161>
	                              <UIApplicationMain+159>
	                               <main+111>
	                                <start+1>
<Error>: CGContextSaveGState: invalid context 0x0. Backtrace:
	  <-[UIBezierPath fill]+26>
	   <-[AveBorderedButton fillViewImage]+358>
	    <-[AveBorderedButton updateFillView]+81>
	     <-[AveBorderedButton updateHighlightedStateAnimated]+43>
	      <-[AveBorderedButton setHighlighted:]+88>
	       <-[AvePurchaseButton setHighlighted:]+173>
	        <-[UITableViewCell _updateHighlightColorsForView:highlighted:]+771>
	         <-[UITableViewCell _updateHighlightColorsForView:highlighted:]+1120>
	          <-[UITableViewCell _updateHighlightColors]+115>
	           <-[UITableViewCell prepareForReuse]+421>
	            <-[UITableView _dequeueReusableViewOfType:withIdentifier:]+1476>
	             <-[BFMasterTableViewController tableView:cellForRowAtIndexPath:]+436>
	              <-[UITableView _createPreparedCellForGlobalRow:withIndexPath:willDisplay:]+757>
	               <-[UITableView _createPreparedCellForGlobalRow:willDisplay:]+74>
	                <-[UITableView _updateVisibleCellsNow:isRecursive:]+3295>
	                 <-[UITableView _performWithCachedTraitCollection:]+110>
	                  <-[UITableView layoutSubviews]+222>
	                   <-[UIView(CALayerDelegate) layoutSublayersOfLayer:]+1237>
	                    <-[CALayer layoutSublayers]+146>
	                     <_ZN2CA5Layer16layout_if_neededEPNS_11TransactionE+366>
	                      <_ZN2CA5Layer28layout_and_display_if_neededEPNS_11TransactionE+24>
	                       <_ZN2CA7Context18commit_transactionEPNS_11TransactionE+280>
	                        <_ZN2CA11Transaction6commitEv+475>
	                         <_ZN2CA11Transaction17observer_callbackEP19__CFRunLoopObservermPv+113>
	                          <__CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__+23>
	                           <__CFRunLoopDoObservers+391>
	                            <__CFRunLoopRun+1198>
	                             <CFRunLoopRunSpecific+420>
	                              <GSEventRunModal+161>
	                               <UIApplicationMain+159>
	                                <main+111>
	                                 <start+1>
<Error>: CGContextSetFlatness: invalid context 0x0. Backtrace:
	  <-[UIBezierPath fill]+47>
	   <-[AveBorderedButton fillViewImage]+358>
	    <-[AveBorderedButton updateFillView]+81>
	     <-[AveBorderedButton updateHighlightedStateAnimated]+43>
	      <-[AveBorderedButton setHighlighted:]+88>
	       <-[AvePurchaseButton setHighlighted:]+173>
	        <-[UITableViewCell _updateHighlightColorsForView:highlighted:]+771>
	         <-[UITableViewCell _updateHighlightColorsForView:highlighted:]+1120>
	          <-[UITableViewCell _updateHighlightColors]+115>
	           <-[UITableViewCell prepareForReuse]+421>
	            <-[UITableView _dequeueReusableViewOfType:withIdentifier:]+1476>
	             <-[BFMasterTableViewController tableView:cellForRowAtIndexPath:]+436>
	              <-[UITableView _createPreparedCellForGlobalRow:withIndexPath:willDisplay:]+757>
	               <-[UITableView _createPreparedCellForGlobalRow:willDisplay:]+74>
	                <-[UITableView _updateVisibleCellsNow:isRecursive:]+3295>
	                 <-[UITableView _performWithCachedTraitCollection:]+110>
	                  <-[UITableView layoutSubviews]+222>
	                   <-[UIView(CALayerDelegate) layoutSublayersOfLayer:]+1237>
	                    <-[CALayer layoutSublayers]+146>
	                     <_ZN2CA5Layer16layout_if_neededEPNS_11TransactionE+366>
	                      <_ZN2CA5Layer28layout_and_display_if_neededEPNS_11TransactionE+24>
	                       <_ZN2CA7Context18commit_transactionEPNS_11TransactionE+280>
	                        <_ZN2CA11Transaction6commitEv+475>
	                         <_ZN2CA11Transaction17observer_callbackEP19__CFRunLoopObservermPv+113>
	                          <__CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__+23>
	                           <__CFRunLoopDoObservers+391>
	                            <__CFRunLoopRun+1198>
	                             <CFRunLoopRunSpecific+420>
	                              <GSEventRunModal+161>
	                               <UIApplicationMain+159>
	                                <main+111>
	                                 <start+1>
<Error>: CGContextAddPath: invalid context 0x0. Backtrace:
	  <-[UIBezierPath fill]+66>
	   <-[AveBorderedButton fillViewImage]+358>
	    <-[AveBorderedButton updateFillView]+81>
	     <-[AveBorderedButton updateHighlightedStateAnimated]+43>
	      <-[AveBorderedButton setHighlighted:]+88>
	       <-[AvePurchaseButton setHighlighted:]+173>
	        <-[UITableViewCell _updateHighlightColorsForView:highlighted:]+771>
	         <-[UITableViewCell _updateHighlightColorsForView:highlighted:]+1120>
	          <-[UITableViewCell _updateHighlightColors]+115>
	           <-[UITableViewCell prepareForReuse]+421>
	            <-[UITableView _dequeueReusableViewOfType:withIdentifier:]+1476>
	             <-[BFMasterTableViewController tableView:cellForRowAtIndexPath:]+436>
	              <-[UITableView _createPreparedCellForGlobalRow:withIndexPath:willDisplay:]+757>
	               <-[UITableView _createPreparedCellForGlobalRow:willDisplay:]+74>
	                <-[UITableView _updateVisibleCellsNow:isRecursive:]+3295>
	                 <-[UITableView _performWithCachedTraitCollection:]+110>
	                  <-[UITableView layoutSubviews]+222>
	                   <-[UIView(CALayerDelegate) layoutSublayersOfLayer:]+1237>
	                    <-[CALayer layoutSublayers]+146>
	                     <_ZN2CA5Layer16layout_if_neededEPNS_11TransactionE+366>
	                      <_ZN2CA5Layer28layout_and_display_if_neededEPNS_11TransactionE+24>
	                       <_ZN2CA7Context18commit_transactionEPNS_11TransactionE+280>
	                        <_ZN2CA11Transaction6commitEv+475>
	                         <_ZN2CA11Transaction17observer_callbackEP19__CFRunLoopObservermPv+113>
	                          <__CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__+23>
	                           <__CFRunLoopDoObservers+391>
	                            <__CFRunLoopRun+1198>
	                             <CFRunLoopRunSpecific+420>
	                              <GSEventRunModal+161>
	                               <UIApplicationMain+159>
	                                <main+111>
	                                 <start+1>
<Error>: CGContextDrawPath: invalid context 0x0. Backtrace:
	  <-[UIBezierPath fill]+98>
	   <-[AveBorderedButton fillViewImage]+358>
	    <-[AveBorderedButton updateFillView]+81>
	     <-[AveBorderedButton updateHighlightedStateAnimated]+43>
	      <-[AveBorderedButton setHighlighted:]+88>
	       <-[AvePurchaseButton setHighlighted:]+173>
	        <-[UITableViewCell _updateHighlightColorsForView:highlighted:]+771>
	         <-[UITableViewCell _updateHighlightColorsForView:highlighted:]+1120>
	          <-[UITableViewCell _updateHighlightColors]+115>
	           <-[UITableViewCell prepareForReuse]+421>
	            <-[UITableView _dequeueReusableViewOfType:withIdentifier:]+1476>
	             <-[BFMasterTableViewController tableView:cellForRowAtIndexPath:]+436>
	              <-[UITableView _createPreparedCellForGlobalRow:withIndexPath:willDisplay:]+757>
	               <-[UITableView _createPreparedCellForGlobalRow:willDisplay:]+74>
	                <-[UITableView _updateVisibleCellsNow:isRecursive:]+3295>
	                 <-[UITableView _performWithCachedTraitCollection:]+110>
	                  <-[UITableView layoutSubviews]+222>
	                   <-[UIView(CALayerDelegate) layoutSublayersOfLayer:]+1237>
	                    <-[CALayer layoutSublayers]+146>
	                     <_ZN2CA5Layer16layout_if_neededEPNS_11TransactionE+366>
	                      <_ZN2CA5Layer28layout_and_display_if_neededEPNS_11TransactionE+24>
	                       <_ZN2CA7Context18commit_transactionEPNS_11TransactionE+280>
	                        <_ZN2CA11Transaction6commitEv+475>
	                         <_ZN2CA11Transaction17observer_callbackEP19__CFRunLoopObservermPv+113>
	                          <__CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__+23>
	                           <__CFRunLoopDoObservers+391>
	                            <__CFRunLoopRun+1198>
	                             <CFRunLoopRunSpecific+420>
	                              <GSEventRunModal+161>
	                               <UIApplicationMain+159>
	                                <main+111>
	                                 <start+1>
<Error>: CGContextRestoreGState: invalid context 0x0. Backtrace:
	  <-[AveBorderedButton fillViewImage]+358>
	   <-[AveBorderedButton updateFillView]+81>
	    <-[AveBorderedButton updateHighlightedStateAnimated]+43>
	     <-[AveBorderedButton setHighlighted:]+88>
	      <-[AvePurchaseButton setHighlighted:]+173>
	       <-[UITableViewCell _updateHighlightColorsForView:highlighted:]+771>
	        <-[UITableViewCell _updateHighlightColorsForView:highlighted:]+1120>
	         <-[UITableViewCell _updateHighlightColors]+115>
	          <-[UITableViewCell prepareForReuse]+421>
	           <-[UITableView _dequeueReusableViewOfType:withIdentifier:]+1476>
	            <-[BFMasterTableViewController tableView:cellForRowAtIndexPath:]+436>
	             <-[UITableView _createPreparedCellForGlobalRow:withIndexPath:willDisplay:]+757>
	              <-[UITableView _createPreparedCellForGlobalRow:willDisplay:]+74>
	               <-[UITableView _updateVisibleCellsNow:isRecursive:]+3295>
	                <-[UITableView _performWithCachedTraitCollection:]+110>
	                 <-[UITableView layoutSubviews]+222>
	                  <-[UIView(CALayerDelegate) layoutSublayersOfLayer:]+1237>
	                   <-[CALayer layoutSublayers]+146>
	                    <_ZN2CA5Layer16layout_if_neededEPNS_11TransactionE+366>
	                     <_ZN2CA5Layer28layout_and_display_if_neededEPNS_11TransactionE+24>
	                      <_ZN2CA7Context18commit_transactionEPNS_11TransactionE+280>
	                       <_ZN2CA11Transaction6commitEv+475>
	                        <_ZN2CA11Transaction17observer_callbackEP19__CFRunLoopObservermPv+113>
	                         <__CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__+23>
	                          <__CFRunLoopDoObservers+391>
	                           <__CFRunLoopRun+1198>
	                            <CFRunLoopRunSpecific+420>
	                             <GSEventRunModal+161>
	                              <UIApplicationMain+159>
	                               <main+111>
	                                <start+1>
<Error>: CGContextSetBlendMode: invalid context 0x0. Backtrace:
	  <-[AveBorderedButton fillViewImage]+376>
	   <-[AveBorderedButton updateFillView]+81>
	    <-[AveBorderedButton updateHighlightedStateAnimated]+43>
	     <-[AveBorderedButton setHighlighted:]+88>
	      <-[AvePurchaseButton setHighlighted:]+173>
	       <-[UITableViewCell _updateHighlightColorsForView:highlighted:]+771>
	        <-[UITableViewCell _updateHighlightColorsForView:highlighted:]+1120>
	         <-[UITableViewCell _updateHighlightColors]+115>
	          <-[UITableViewCell prepareForReuse]+421>
	           <-[UITableView _dequeueReusableViewOfType:withIdentifier:]+1476>
	            <-[BFMasterTableViewController tableView:cellForRowAtIndexPath:]+436>
	             <-[UITableView _createPreparedCellForGlobalRow:withIndexPath:willDisplay:]+757>
	              <-[UITableView _createPreparedCellForGlobalRow:willDisplay:]+74>
	               <-[UITableView _updateVisibleCellsNow:isRecursive:]+3295>
	                <-[UITableView _performWithCachedTraitCollection:]+110>
	                 <-[UITableView layoutSubviews]+222>
	                  <-[UIView(CALayerDelegate) layoutSublayersOfLayer:]+1237>
	                   <-[CALayer layoutSublayers]+146>
	                    <_ZN2CA5Layer16layout_if_neededEPNS_11TransactionE+366>
	                     <_ZN2CA5Layer28layout_and_display_if_neededEPNS_11TransactionE+24>
	                      <_ZN2CA7Context18commit_transactionEPNS_11TransactionE+280>
	                       <_ZN2CA11Transaction6commitEv+475>
	                        <_ZN2CA11Transaction17observer_callbackEP19__CFRunLoopObservermPv+113>
	                         <__CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__+23>
	                          <__CFRunLoopDoObservers+391>
	                           <__CFRunLoopRun+1198>
	                            <CFRunLoopRunSpecific+420>
	                             <GSEventRunModal+161>
	                              <UIApplicationMain+159>
	                               <main+111>
	                                <start+1>
```

**Comment by ReDetection** [comment]
@AndreasVerhoeven please file another issue, this one is narrowed to changing `state` property. Also would be nice not to blow message to such a huge size – you should leave only one backtrace (anyway it's repeated because of tableview), but if you really want to keep them all – you'd rather [attach a gist](https://gist.github.com)

**Comment by AndreasVerhoeven** [comment]
@jonbrown21 's issue has been fixed in 1.0.6

**Comment by jonbrown21** [comment]
Sorry about that, I will note that for next time.
